File: Java/Medium/Accepted/shortest-path-with-alternating-colors_3.java
class Solution {
    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        
        int[] result = new int[n];
        Arrays.fill(result, -1);
        result[0] = 0;
        Map<Integer, List<Integer>> redMap = new HashMap<>();
        Map<Integer, List<Integer>> blueMap = new HashMap<>();
        for (int[] edge : redEdges) {
            redMap.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
        }
        for (int[] edge : blueEdges) {
            blueMap.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
        }
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0});
        queue.offer(new int[]{0, 1});
        int step = 0;
        Set<String> visited = new HashSet<>();
        visited.add(0 + " " + 0);
        visited.add(0 + " " + 1);
        while (!queue.isEmpty()) {
            int size = queue.size();
            step++;
            for (int i = 0; i < size; i++) {
                int[] node = queue.poll();
                int color = node[1];
                int nextColor = color == 0 ? 1 : 0;
                Map<Integer, List<Integer>> map = color == 0 ? redMap : blueMap;
                if (map.containsKey(node[0])) {
                    for (int next : map.get(node[0])) {
                        if (result[next] == -1) {
                            result[next] = step;
                        }
                        if (!visited.contains(next + " " + nextColor)) {
                            visited.add(next + " " + nextColor);
                            queue.offer(new int[]{next, nextColor});
                        }
                    }
                }
            }
        }
        return result;
    }
}

File: C++/Medium/Accepted/shortest-path-with-alternating-colors_3.cpp
class Solution {
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>>& redEdges, vector<vector<int>>& blueEdges) {
        
        vector<vector<pair<int, int>>> graph(n);
        for (auto e : redEdges) graph[e[0]].push_back({e[1], 0});
        for (auto e : blueEdges) graph[e[0]].push_back({e[1], 1});
        vector<vector<int>> dist(n, vector<int>(2, INT_MAX));
        dist[0][0] = 0;
        dist[0][1] = 0;
        queue<pair<int, int>> q;
        q.push({0, 0});
        q.push({0, 1});
        while (!q.empty()) {
            auto [u, c] = q.front();
            q.pop();
            for (auto [v, nc] : graph[u]) {
                if (nc == c) continue;
                if (dist[v][nc] > dist[u][c] + 1) {
                    dist[v][nc] = dist[u][c] + 1;
                    q.push({v, nc});
                }
            }
        }
        vector<int> ans(n, INT_MAX);
        for (int i = 0; i < n; i++) {
            ans[i] = min(dist[i][0], dist[i][1]);
            if (ans[i] == INT_MAX) ans[i] = -1;
        }
        return ans;
    }
};